include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,376 @@
+ */
+
+#ifndef _ONEDPL_SCAN_BY_SEGMENT_IMPL_H
+#define _ONEDPL_SCAN_BY_SEGMENT_IMPL_H
+
+#if _ONEDPL_BACKEND_SYCL
+
+#    include <oneapi/dpl/pstl/algorithm_fwd.h>
+#    include <oneapi/dpl/pstl/parallel_backend.h>
+#    include <oneapi/dpl/pstl/hetero/utils_hetero.h>
+
+#    include <oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h>
+#    include <oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h>
+#    include <oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h>
+
+#    include <array>
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace internal
+{
+
+// This function is responsible for performing an exclusive segmented scan between work items in shared local memory.
+// Each work item passes __delta_local_id which is the distance to the closest lower indexed work item that
+// processes a segment end. A SLM accessor that is twice the work-group size must be provided.
+// Example across eight work items with addition as the binary operation :
+// -----------------------------
+// Local ID    : 0 1 2 3 4 5 6 7
+// Accumulator : 2 3 1 3 0 5 2 1
+// Delta       : 0 1 2 3 0 1 0 0
+// -----------------------------
+// Return Vals : 0 2 5 6 9 0 5 2
+// -----------------------------
+template <typename _NdItem, typename _LocalAcc, typename _IdxType, typename _ValueType, typename _BinaryOp>
+inline _ValueType
+wg_segmented_scan(_NdItem __item, _LocalAcc __local_acc, _IdxType __local_id, _IdxType __delta_local_id,
+                  _ValueType __accumulator, _ValueType __identity, _BinaryOp __binary_op, ::std::size_t __wgroup_size)
+{
+    _IdxType __first = 0;
+    __local_acc[__local_id] = __accumulator;
+
+    __dpl_sycl::__group_barrier(__item);
+
+    for (::std::size_t __i = 1; __i < __wgroup_size; __i *= 2)
+    {
+        if (__delta_local_id >= __i)
+            __accumulator = __binary_op(__local_acc[__first + __local_id - __i], __accumulator);
+
+        __first = __wgroup_size - __first;
+        __local_acc[__first + __local_id] = __accumulator;
+        __dpl_sycl::__group_barrier(__item);
+    }
+
+    return (__local_id ? __local_acc[__first + __local_id - 1] : __identity);
+}
+
+enum class scan_type
+{
+    inclusive,
+    exclusive
+};
+
+template <scan_type __scan_type, typename... Name>
+class __sycl_segmented_scan_kernel1;
+
+template <scan_type __scan_type, typename... Name>
+class __sycl_segmented_scan_kernel2;
+
+template <scan_type __scan_type>
+struct sycl_scan_by_segment_impl
+{
+    template <typename... _Name>
+    using _KernelName1 = __sycl_segmented_scan_kernel1<__scan_type, _Name...>;
+
+    template <typename... _Name>
+    using _KernelName2 = __sycl_segmented_scan_kernel2<__scan_type, _Name...>;
+
+    template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3,
+              typename _BinaryPredicate, typename _BinaryOperator, typename _T>
+    void
+    sycl_scan_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values, _Range3&& __out_values,
```

Comment:
What the point to have is as a separate method?

Reply:
I have looked at this and agree it is unnecessary. I have removed this function and simply added the implementation in the function call operator.